File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    constructor() {
        this.max = 0
    }

    diameterOfBinaryTree(root) {
        this.dfs(root)
        return this.max
    }

    dfs(node) {
        if(node === null) return 0
        let left = this.dfs(node.left)
        let right = this.dfs(node.right)
        
        this.max = Math.max(this.max, left + right)  // track diameter
        
        return 1 + Math.max(left, right)  // return depth to parent
    }
}

File: Data Structures & Algorithms/replace-elements-with-greatest-element-on-right-side/submission-0.java
class Solution {
    public int[] replaceElements(int[] arr) {
        for(int i=0; i<arr.length-1; i++) {
            List<Integer> track = new ArrayList<>();
            for(int j=i+1; j<arr.length; j++) {
                track.add(arr[j]);
            }
            int max = Integer.MIN_VALUE;
            for(int num : track) {
                max = Math.max(num, max);
            }
            arr[i] = max;
        }
        arr[arr.length-1] = -1;
        return arr;
    }
}